### ISSUE TITLE
"Undefined local variable" error when refering to it in a constant

### ISSUE THREAD

**Issue by pyrsmk** [crystal-lang/crystal#11272]
## Bug Report

This code fails :

```cr
path = Path["dir1/dir2/file1.csv"]
FILENAME = "#{path.parent}/#{path.basename[0..-4]}.updated.csv"
puts FILENAME
```

With the following error : 

```
In test.cr:7:15

 7 | FILENAME = "#{path.parent}/#{path.basename[0..-4]}.updated.csv"
                   ^---
Error: undefined local variable or method 'path' for top-level
```

Modifying the constant by a regular variable works as expected : 

```cr
path = Path["dir1/dir2/file1.csv"]
filename = "#{path.parent}/#{path.basename[0..-4]}.updated.csv"
puts filename
```

---

Here's my environment : 

```
❯ crystal -v
Crystal 1.1.1 [6d9a1d583] (2021-07-26)

LLVM: 10.0.1
Default target: x86_64-unknown-linux-gnu
```

**Comment by straight-shoota** [comment]
What would be a bug here? The error is correct and expected. Constant's can't reference local variables.

**Comment by pyrsmk** [comment]
Hum ok. Sorry, it comes from my background of dynamic languages where you can do whatever you want 😅.

Maybe the error could be more user friendly then ? Because it's kind of unclear IMO.

**Comment by beta-ziliani** [comment]
> Maybe the error could be more user friendly then ?

Agreed. As a bit of a longer explanation, constants are values computed statically during compilation and that's why they can't refer to variables.

**Comment by Blacksmoke16** [comment]
For reference you could do this:

```cr
FILENAME = begin
  path = Path["dir1/dir2/file1.csv"]
  "#{path.parent}/#{path.basename[0..-4]}.updated.csv"
end

puts FILENAME
```

**Comment by jgaskins** [comment]
> What would be a bug here? The error is correct and expected. Constant's values can't reference local variables.

Is this restriction documented? I didn't see it in [the "Constants" page in the docs website](https://crystal-lang.org/reference/syntax_and_semantics/constants.html).

I was under the same impression that @pyrsmk was, that it was just assigning a value to an immutable binding that was accessible from anywhere in the app. I ran into this for the first time this evening and thought I was _losing my mind_.

**Comment by j8r** [comment]
Yeah, maybe pointing to a doc link, which has the explanations of what a constant is and what can be done and not, like here they can't depend on runtime values.
